Don't add rpath to swift with statically linking

keith opened the pull request:
This updates the rpath linking logic to only add the rpath that points
to the swift dylibs in the case that the libraries are not statically
linked into the binary.

This fix is marginally relevant to https://bugs.swift.org/browse/SR-1967, but does not solve the core problem here.

dduan commented:
@swift-ci please test

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1283,10 +1284,10 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
   }
 
-  // FIXME: We probably shouldn't be adding an rpath here unless we know ahead
-  // of time the standard library won't be copied.
```
Please preserve this FIXME in the new code as well—it's referencing the rest of SR-1967.

keith replied:
Done, also moved this up into the else above

jrose-apple reviewed (commented):
Thanks for picking this up, Keith! Can you add a test to test/Driver/linker.swift as well?

keith committed:
Don't add rpath to swift with statically linking

This updates the rpath linking logic to only add the rpath that points
to the swift dylibs in the case that the libraries are not statically
linked into the binary.

keith commented:
Yep I'll look at adding that test!

keith commented:
(sorry about the tests, the first push here was bogus)

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test

keith committed:
Add test for static-stdlib

This test validates the arguments passed to the linker when statically
linking the swift standard library. Currently in order to ensure that no
-rpath is passed, we're using FileCheck's -implicit-check-not flag, and
strictly validating the order of the arguments. The order doesn't really
matter here but is required for that flag to validate that no -rpath is
passed.

keith commented:
I've just pushed a test for this. I would love some feedback on how I'm using `FileCheck` to validated the absence of `-rpath`. AFACIT there's no good way to do this. Really I want something like:

```swift
// SWIFT_STATIC-DAG-NOT: -rpath
```

Instead I'm using the `-implicit-check-not` flag, which works in this case it seems, but requires a more ordered list of arguments, rather than ignoring the order and using `-DAG` for all of these (since the order doesn't actually matter)

dduan commented:
@swift-ci please test

dduan commented:
@swift-ci please test

keith commented:
@swift-ci please test

dduan commented:
@swift-ci please test

keith commented:
Looks like these failures are because I force pushed, it's still building the new sha.

dduan commented:
@swift-ci please test

keith commented:
(it seems to still be running)

dduan commented:
@swift-ci please smoke test OS X Platform

keith commented:
Thanks!
